### PR TITLE
feat: add batching for primitives and instance meshes for V9 output

### PR DESCRIPTION
**Pull request by christjt**

### CODE REVIEW

File: viewer/packages/cad-parsers/src/cad/types.ts [hunk 1]
```diff
@@ -48,6 +49,11 @@ export interface ConsumedSector {
   levelOfDetail: LevelOfDetail;
   group: AutoDisposeGroup | undefined;
   instancedMeshes: InstancedMeshFile[] | undefined;
+  geometryBatchingQueue?: [
+    type: RevealGeometryCollectionType,
+    geometryBuffer: THREE.BufferGeometry,
+    instanceId: string | undefined
```
Comment (larsmoa) [inline viewer/packages/cad-parsers/src/cad/types.ts:55]: I think this deserves its own type.

File: viewer/packages/rendering/src/rendering/primitives.ts [hunk 2]
```diff
@@ -361,8 +371,18 @@ function createGeneralRings(
...
 
-  mesh.onBeforeRender = () => updateMaterialInverseModelMatrix(material, mesh.matrixWorld);
+  mesh.onBeforeRender = (_0, _1, camera: THREE.Camera) => {
+    updateMaterialInverseModelMatrix(material, mesh.matrixWorld);
+
+    (material.uniforms.modelMatrix?.value as THREE.Matrix4)?.copy(mesh.matrixWorld);
+    (material.uniforms.viewMatrix?.value as THREE.Matrix4)?.copy(camera.matrixWorld).invert();
+    (material.uniforms.projectionMatrix?.value as THREE.Matrix4)?.copy(camera.projectionMatrix);
+    (material.uniforms.normalMatrix?.value as THREE.Matrix3)?.copy(mesh.normalMatrix);
+    (material.uniforms.cameraPosition?.value as THREE.Vector3)?.copy(camera.position);
```
Comment (larsmoa) [inline viewer/packages/rendering/src/rendering/primitives.ts:381]: Duplicated code from the snippet above, create a function for it.

File: viewer/packages/rendering/src/sector/CadNode.ts [hunk 3]
```diff
@@ -177,6 +184,21 @@ export class CadNode extends THREE.Object3D {
     this._instancedMeshManager.removeSectorInstancedMeshes(sectorId);
   }
 
+  public batchGeometry(
+    geometryBatchingQueue: [
+      type: RevealGeometryCollectionType,
+      geometryBuffer: THREE.BufferGeometry,
+      instanceId: string | undefined
```
Comment (larsmoa) [inline viewer/packages/rendering/src/sector/CadNode.ts:191]: Create type for this (mentioned above).

File: viewer/packages/rendering/src/sector/CadNode.ts [hunk 4]
```diff
@@ -177,6 +184,21 @@ export class CadNode extends THREE.Object3D {
...
+    ][],
+    sectorId: number
+  ) {
```
Comment (larsmoa) [inline viewer/packages/rendering/src/sector/CadNode.ts:194]: Always add return type of public functions (added as a lint step that soon will hit this)

File: viewer/packages/rendering/src/sector/CadNode.ts [hunk 5]
```diff
@@ -177,6 +184,21 @@ export class CadNode extends THREE.Object3D {
...
+    this._geometryBatchingManager.batchGeometries(geometryBatchingQueue, sectorId);
+  }
+
+  public removeBatchedSectorGeometries(sectorId: number) {
```
Comment (larsmoa) [inline viewer/packages/rendering/src/sector/CadNode.ts:198]: Add return type

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 6]
```diff
@@ -70,7 +74,17 @@ export class GltfSectorParser {
...
-  private processNode(node: Node, glbHeaderData: GlbHeaderData, data: ArrayBuffer) {
+  private processNode(
+    node: Node,
+    glbHeaderData: GlbHeaderData,
+    data: ArrayBuffer
+  ):
+    | {
+        type: RevealGeometryCollectionType;
+        buffer: THREE.InstancedBufferGeometry | THREE.BufferGeometry;
+        instanceId?: number;
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:85]: Create type for this (mentioned above).

**Review by larsmoa — CHANGES_REQUESTED** [review]

File: viewer/packages/sector-parser/src/types.ts [hunk 7]
```diff
@@ -95,13 +95,8 @@ export type GeometryProcessingPayload = {
...
-  | Uint16ArrayConstructor
-  | Int32ArrayConstructor
-  | Uint32ArrayConstructor
-  | Uint8ClampedArrayConstructor
-  | Float32ArrayConstructor
-  | Float64ArrayConstructor;
+export type ParsedGeometry = {
+  type: RevealGeometryCollectionType;
+  geometryBuffer: THREE.BufferGeometry;
+  instanceId?: string;
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/types.ts:101]: Why is this `string` somewhere and `number` elsewhere?

**Review by larsmoa — APPROVED** [review]

File: viewer/packages/sector-parser/src/types.ts (hunk 7 above)
Comment (christjt) [reply viewer/packages/sector-parser/src/types.ts:101]: Should be string consequently everywhere now

**Pull request merged**